lib/db_connection/watcher.ex

```diff
@@ -1,29 +1,56 @@
 defmodule DBConnection.Watcher do
   @moduledoc false
+  @name __MODULE__
 
   use GenServer
 
-  def start_link(owner) do
-    GenServer.start_link(__MODULE__, owner)
+  def start_link do
+    GenServer.start_link(__MODULE__, :ok, name: @name)
   end
 
-  def init(owner) do
-    ref = Process.monitor(owner)
-    try do
-      Process.link(owner)
-    catch
-      :error, :noproc ->
-        {:stop, {:shutdown, :noproc}}
-    else
-      _ ->
-        {:ok, ref}
+  def watch(supervisor, args) when is_list(args) do
+    GenServer.call(@name, {:watch, supervisor, args}, :infinity)
+  end
+
+  def init(:ok) do
+    Process.flag(:trap_exit, true)
+    {:ok, {%{}, %{}}}
+  end
+
+  def handle_call({:watch, supervisor, args}, {caller_pid, _ref}, {caller_refs, started_refs}) do
+    case Supervisor.start_child(supervisor, args) do
+      {:ok, started_pid} ->
+        Process.link(caller_pid)
+        caller_ref = Process.monitor(caller_pid)
+        started_ref = Process.monitor(started_pid)
+        caller_refs = Map.put(caller_refs, caller_ref, {supervisor, started_pid, started_ref})
+        started_refs = Map.put(started_refs, started_ref, {caller_pid, caller_ref})
+        {:reply, {:ok, started_pid}, {caller_refs, started_refs}}
+      other ->
+        {:reply, other, {caller_refs, started_refs}}
     end
   end
 
-  def handle_info({:DOWN, ref, _, _, reason}, ref) do
-    {:stop, {:shutdown, reason}, ref}
+  def handle_info({:DOWN, ref, _, _, _}, {caller_refs, started_refs}) do
+    case caller_refs do
+      %{^ref => {supervisor, started_pid, started_ref}} ->
+        Process.demonitor(started_ref, [:flush])
+        Supervisor.terminate_child(supervisor, started_pid)
+        {:noreply, {Map.delete(caller_refs, ref), Map.delete(started_refs, started_ref)}}
+      %{} ->
+        case started_refs do
+          %{^ref => {caller_pid, caller_ref}} ->
+            Process.demonitor(caller_ref, [:flush])
+            Process.unlink(caller_pid)
```

Comment:
Lets skip the `unlink` because we trap exits and handle exit signals, and technically watcher could be killed by its parent supervisor between `unlink` and `kill`.